cirq/qis/states.py

```diff
@@ -36,6 +37,503 @@
+        data = one_hot(index=state, shape=(dim,), dtype=dtype)
+    else:
+        data = np.array(state, copy=False)
+        if qid_shape is None:
+            qid_shape = infer_qid_shape(state)
+        if data.ndim == 1:
+            if len(qid_shape) == np.prod(qid_shape, dtype=int) and data.dtype.kind != 'c':
+                raise ValueError(
+                    'Because len(qid_shape) == product(qid_shape), it is '
+                    'ambiguous whether the given state contains '
+                    'state vector amplitudes or per-qudit computational basis '
+                    'values. In this situation you are required to pass '
+                    'in a state vector that is a numpy array with a complex '
+                    'dtype.'
+                )
+            if data.shape == (len(qid_shape),):
+                # array contains per-qudit computational basis values
+                data = _qudit_values_to_state_tensor(
+                    state_vector=data, qid_shape=qid_shape, dtype=dtype
+                )
+        if copy or dtype and dtype != data.dtype:
+            if dtype and dtype != data.dtype:
+                data = data.astype(dtype, casting='unsafe', copy=True)
+            else:
+                data = data.copy()
+    return QuantumState(data=data, qid_shape=qid_shape, validate=validate, dtype=dtype, atol=atol)
+
+
+def density_matrix(
+    state: np.ndarray,
+    qid_shape: Optional[Tuple[int, ...]] = None,
+    *,  # Force keyword arguments
+    copy: bool = False,
+    validate: bool = True,
+    dtype: Optional[Type[np.number]] = None,
+    atol: float = 1e-7,
+) -> QuantumState:
+    """Create a QuantumState object from a density matrix.
+
+    Args:
+        state: The density matrix.
+        qid_shape: The qid shape.
+        copy: Whether to copy the density matrix.
+        validate: Whether to check if the given data and qid shape
+            represent a valid quantum state with the given dtype.
+        dtype: The expected data type.
+        atol: Absolute numerical tolerance to use for validation.
+
+    Raises:
+        ValueError: Invalid density matrix.
+    """
+    if state.ndim != 2 or state.shape[0] != state.shape[1]:
+        raise ValueError('A density matrix must be a square matrix. ' f'Got shape {state.shape}.')
+    dim, _ = state.shape
+    if qid_shape is None:
+        qid_shape = _infer_qid_shape_from_dimension(dim)
+    return QuantumState(
+        data=state.copy() if copy else state,
+        qid_shape=qid_shape,
+        dtype=dtype,
+        validate=validate,
+        atol=atol,
+    )
+
+
+def _infer_qid_shape_from_dimension(dim: int) -> Tuple[int, ...]:
+    if dim != 0 and dim & dim - 1 == 0:
+        # dim is a power of 2, assume qubits
+        n_qubits = dim.bit_length() - 1
+        return (2,) * n_qubits
+    # dim is not a power of 2, assume a single qudit
+    return (dim,)
+
+
+def infer_qid_shape(*states: 'cirq.QUANTUM_STATE_LIKE') -> Tuple[int, ...]:
```

Comment:
can we split qid shape inference to another PR? We can throw an error in the meantime from `quantum_state()`

Reply:
Yes, that makes sense. I've removed it from this PR and will open a separate one.